nilearn/utils/discovery.py

```diff
@@ -0,0 +1,188 @@
+"""Utilities to discover nilearn objects."""
+
+import inspect
+import pkgutil
+from importlib import import_module
+from operator import itemgetter
+from pathlib import Path
+
+from sklearn.base import BaseEstimator, ClusterMixin, TransformerMixin
+from sklearn.utils._testing import ignore_warnings
+
+from nilearn._utils.param_validation import check_parameter_in_allowed
+
+_MODULE_TO_IGNORE = {
+    "_utils",
+    "conftest",
+    "externals",
+    "input_data",
+    "tests",
+}
+
+
+def _skip_module(module_name: str):
+    module_parts = module_name.split(".")
+    return bool(
+        any(part in _MODULE_TO_IGNORE for part in module_parts)
+        or "._" in module_name
+    )
+
+
+def all_estimators(type_filter=None):
+    """Get a list of all estimators from `nilearn`.
+
+    This function crawls the module and gets all classes that inherit
+    from sklearn BaseEstimator.
+    Classes that are defined in test-modules are not included.
+
+    Parameters
+    ----------
+    type_filter : {"classifier",  "cluster", "regressor", "masker", \
+                  "multi_masker", "transformer"} \
+                  or list of such strings, default=None
+        Which kind of estimators should be returned.
+        If ``None``, no filter is applied and all estimators are returned.
+        Possible values are
+        "classifier", "cluster", "regressor", "masker",
+        "multi_masker", "transformer"
+        to get estimators only of these specific types,
+        or a list of these to get the estimators
+        that fit at least one of the types.
+
+    Returns
+    -------
+    estimators : functions : :obj:`list` of :obj:`tuple`
```

Comment:
I could not understand why we have `functions :` here.

Reply:
copy paste error